File: Components/DevExpressVCL/Demos/ExpressQuantumGrid/CBuilder/LayoutViewCarouselModeDemo/LayoutViewCarouselModeDemoMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "LayoutViewCarouselModeDemoMain.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxClasses"
#pragma link "cxControls"
#pragma link "cxCustomData"
#pragma link "cxData"
#pragma link "cxDataStorage"
#pragma link "cxDBData"
#pragma link "cxEdit"
#pragma link "cxFilter"
#pragma link "cxGraphics"
#pragma link "cxGrid"
#pragma link "cxGridCustomLayoutView"
#pragma link "cxGridCustomTableView"
#pragma link "cxGridCustomView"
#pragma link "cxGridDBLayoutView"
#pragma link "cxGridLayoutView"
#pragma link "cxGridLevel"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxLookAndFeels"
#pragma link "cxNavigator"
#pragma link "cxStyles"
#pragma link "dxLayoutContainer"
#pragma link "dxLayoutLookAndFeels"
#pragma link "dxmdaset"
#pragma link "cxButtons"
#pragma link "cxCheckBox"
#pragma link "cxContainer"
#pragma link "cxDropDownEdit"
#pragma link "cxGridCardView"
#pragma link "cxGridTableView"
#pragma link "cxMaskEdit"
#pragma link "cxTextEdit"
#pragma link "cxTrackBar"
#pragma link "dxLayoutControlAdapters"
#pragma link "dxLayoutcxEditAdapters"
#pragma link "dxToggleSwitch"
#pragma resource "*.dfm"
TfrmMain *frmMain;
//---------------------------------------------------------------------------
__fastcall TfrmMain::TfrmMain(TComponent* Owner)
	: TfmBaseForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TfrmMain::miCustomizeClick(TObject *Sender)
{
	LayoutView->Controller->Customization = True;
}
//---------------------------------------------------------------------------

void __fastcall TfrmMain::btnCustomizeClick(TObject *Sender)
{
	LayoutView->Controller->Customization = True;
}
//---------------------------------------------------------------------------

void __fastcall TfrmMain::cbExpandableRecordsClick(TObject *Sender)
{
	LayoutView->OptionsCustomize->RecordExpanding = cbExpandableRecords->Checked;
}
//---------------------------------------------------------------------------

void __fastcall TfrmMain::cbMultiSelectRecordsClick(TObject *Sender)
{
	LayoutView->OptionsSelection->MultiSelect = cbMultiSelectRecords->Checked;
}
//---------------------------------------------------------------------------

void __fastcall TfrmMain::cbRecordCaptionsClick(TObject *Sender)
{
	LayoutView->OptionsView->RecordCaption->Visible = cbRecordCaptions->Checked;
}
//---------------------------------------------------------------------------

String __fastcall GetDir(String const AFileName, Integer ALevelUp)
{
	Integer I;
	String rv = AFileName;
	for (I = 1; I <= ALevelUp; I++) {
		rv = ExtractFileDir(rv);
	}
	return rv;
}

void __fastcall TfrmMain::FormCreate(TObject *Sender)
{
  FLockCount++;
  try {
	tsAutoPitchAngle->Checked = LayoutView->OptionsView->CarouselMode->AutoPitchAngle;
	cbInterpolationMode->ItemIndex = (int)LayoutView->OptionsView->CarouselMode->InterpolationMode;
	tbPitchAngle->Position = (int)LayoutView->OptionsView->CarouselMode->PitchAngle;
	tbRollAngle->Position = (int)LayoutView->OptionsView->CarouselMode->RollAngle;
	tbRecordCount->Position = LayoutView->OptionsView->CarouselMode->RecordCount;
	tbBackgroundAlphaLevel->Position = LayoutView->OptionsView->CarouselMode->BackgroundRecordAlphaLevel;
	tbStartRecordScale->Position = LayoutView->OptionsView->CarouselMode->BackgroundRecordStartScale;
	tbEndRecordScale->Position = LayoutView->OptionsView->CarouselMode->BackgroundRecordEndScale;
  }
  __finally {
	FLockCount--;
  }
  mdHomes->LoadFromBinaryFile(GetDir(Application->ExeName, 3) + "\\Data\\Homes.dat");
}
//---------------------------------------------------------------------------

void __fastcall TfrmMain::CarouselModePropertiesChange(TObject *Sender)
{
  if (FLockCount) {
	  return ;
  }
  FLockCount++;
  try{
	LayoutView->OptionsView->CarouselMode->AutoPitchAngle = tsAutoPitchAngle->Checked;
	LayoutView->OptionsView->CarouselMode->InterpolationMode = (TdxGPInterpolationMode)(cbInterpolationMode->ItemIndex);
	LayoutView->OptionsView->CarouselMode->PitchAngle = tbPitchAngle->Position;
	LayoutView->OptionsView->CarouselMode->RollAngle = tbRollAngle->Position;
	LayoutView->OptionsView->CarouselMode->RecordCount = tbRecordCount->Position;
	LayoutView->OptionsView->CarouselMode->BackgroundRecordAlphaLevel = tbBackgroundAlphaLevel->Position;
	LayoutView->OptionsView->CarouselMode->BackgroundRecordStartScale = tbStartRecordScale->Position;
	LayoutView->OptionsView->CarouselMode->BackgroundRecordEndScale = tbEndRecordScale->Position;
  }
  __finally {
	FLockCount--;
  }
  liPitchAngle->Enabled = !tsAutoPitchAngle->Checked;
}
//---------------------------------------------------------------------------

void __fastcall TfrmMain::FormShow(TObject *Sender)
{
  LayoutView->DataController->RecNo = (int)(LayoutView->DataController->RecordCount / 2);
}
//---------------------------------------------------------------------------

File: Components/DevExpressVCL/Demos/ExpressQuantumGrid/Delphi/LayoutViewCarouselModeDemo/LayoutViewCarouselModeDemoMain.pas
unit LayoutViewCarouselModeDemoMain;

interface

{$I cxVer.inc}

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms,
  Dialogs, cxStyles, cxCustomData, cxGraphics, cxFilter, cxData, cxDataStorage,
  cxEdit, DB, cxDBData, cxControls, cxGridCustomView, cxGridCustomTableView,
  cxClasses, cxGridLevel, cxGrid, StdCtrls, Menus, cxMemo, cxImage, cxCurrencyEdit,
  cxHyperLinkEdit, cxTextEdit, cxEditRepositoryItems, cxLookAndFeels, cxLookAndFeelPainters,
  dxLayoutContainer, cxGridLayoutView, cxGridDBLayoutView, cxGridCustomLayoutView,
  cxContainer, cxGroupBox, dxLayoutLookAndFeels, ExtCtrls, cxButtons, 
  dxmdaset, BaseForm, cxGridTableView, cxRadioGroup, cxCheckBox, cxGridCardView,
  ComCtrls, ImgList, cxLabel, cxMaskEdit, cxDropDownEdit, cxNavigator,
  dxLayoutControlAdapters, dxLayoutcxEditAdapters, dxLayoutControl,
  cxTrackBar, DBClient, cxExtEditRepositoryItems, dxToggleSwitch;

type
  TfrmMain = class(TfmBaseForm)
    dsHomes: TDataSource;
    miView: TMenuItem;
    EditRepository: TcxEditRepository;
    EditRepositoryImage: TcxEditRepositoryImageItem;
    EditRepositoryMemo: TcxEditRepositoryMemoItem;
    EditRepositoryPrice: TcxEditRepositoryCurrencyItem;
    miCustomize: TMenuItem;
    btnCustomize: TcxButton;
    cbMultiSelectRecords: TcxCheckBox;
    cbRecordCaptions: TcxCheckBox;
    cbExpandableRecords: TcxCheckBox;
    mdHomes: TdxMemData;
    lcMainGroup_Root: TdxLayoutGroup;
    lcMain: TdxLayoutControl;
    lcMainItem1: TdxLayoutItem;
    lcMainItem2: TdxLayoutItem;
    lcMainItem3: TdxLayoutItem;
    lcMainItem4: TdxLayoutItem;
    dxLayoutLookAndFeelList1: TdxLayoutLookAndFeelList;
    dxLayoutSkinLookAndFeel1: TdxLayoutSkinLookAndFeel;
    lcMainGroup1: TdxLayoutGroup;
    tbPitchAngle: TcxTrackBar;
    liPitchAngle: TdxLayoutItem;
    tbEndRecordScale: TcxTrackBar;
    tbBackgroundAlphaLevel: TcxTrackBar;
    lcMainItem6: TdxLayoutItem;
    tbStartRecordScale: TcxTrackBar;
    lcMainItem7: TdxLayoutItem;
    tbRollAngle: TcxTrackBar;
    lcMainItem8: TdxLayoutItem;
    lcMainGroup3: TdxLayoutGroup;
    lcMainItem9: TdxLayoutItem;
    lcMainItem10: TdxLayoutItem;
    tbRecordCount: TcxTrackBar;
    cbInterpolationMode: TcxComboBox;
    lcMainItem11: TdxLayoutItem;
    lcMainGroup7: TdxLayoutGroup;
    Grid: TcxGrid;
    LayoutView: TcxGridDBLayoutView;
    dxLayoutGroup1: TdxLayoutGroup;
    GridLevel1: TcxGridLevel;
    lcMainGroup5: TdxLayoutGroup;
    mdHomesAddress: TMemoField;
    mdHomesBeds: TSmallintField;
    mdHomesBaths: TSmallintField;
    mdHomesHouseSize: TFloatField;
    mdHomesPrice: TFloatField;
    mdHomesFeatures: TMemoField;
    mdHomesYearBuilt: TMemoField;
    mdHomesPhoto: TBlobField;
    LayoutViewLayoutItem1: TcxGridLayoutItem;
    LayoutViewRecId: TcxGridDBLayoutViewItem;
    LayoutViewLayoutItem3: TcxGridLayoutItem;
    LayoutViewAddress: TcxGridDBLayoutViewItem;
    LayoutViewLayoutItem4: TcxGridLayoutItem;
    LayoutViewBeds: TcxGridDBLayoutViewItem;
    LayoutViewLayoutItem5: TcxGridLayoutItem;
    LayoutViewBaths: TcxGridDBLayoutViewItem;
    LayoutViewLayoutItem6: TcxGridLayoutItem;
    LayoutViewHouseSize: TcxGridDBLayoutViewItem;
    LayoutViewLayoutItem8: TcxGridLayoutItem;
    LayoutViewPrice: TcxGridDBLayoutViewItem;
    LayoutViewLayoutItem9: TcxGridLayoutItem;
    LayoutViewFeatures: TcxGridDBLayoutViewItem;
    LayoutViewLayoutItem10: TcxGridLayoutItem;
    LayoutViewYearBuilt: TcxGridDBLayoutViewItem;
    LayoutViewLayoutItem13: TcxGridLayoutItem;
    LayoutViewPhoto: TcxGridDBLayoutViewItem;
    LayoutViewGroup4: TdxLayoutGroup;
    LayoutViewGroup13: TdxLayoutGroup;
    EditRepositorySpinItem: TcxEditRepositorySpinItem;
    LayoutViewGroup3: TdxLayoutGroup;
    tsAutoPitchAngle: TdxToggleSwitch;
    lcMainItem5: TdxLayoutItem;
    procedure miCustomizeClick(Sender: TObject);
    procedure btnCustomizeClick(Sender: TObject);
    procedure cbMultiSelectRecordsClick(Sender: TObject);
    procedure cbExpandableRecordsClick(Sender: TObject);
    procedure cbRecordCaptionsClick(Sender: TObject);
    procedure FormCreate(Sender: TObject);
    procedure CarouselModePropertiesChange(Sender: TObject);
    procedure FormShow(Sender: TObject);
  private
    FLockCount: Integer;
  public
  {$IFDEF EXPRESSBARS}
    procedure PlaceControls; override;
  {$ENDIF}
  end;

var
  frmMain: TfrmMain;

implementation

{$R *.dfm}

uses
  dxGDIPlusClasses;

procedure TfrmMain.miCustomizeClick(Sender: TObject);
begin
  LayoutView.Controller.Customization := True;
end;

procedure TfrmMain.btnCustomizeClick(Sender: TObject);
begin
  LayoutView.Controller.Customization := True;  
end;

procedure TfrmMain.cbExpandableRecordsClick(Sender: TObject);
begin
  LayoutView.OptionsCustomize.RecordExpanding := cbExpandableRecords.Checked;
end;

procedure TfrmMain.cbMultiSelectRecordsClick(Sender: TObject);
begin
  LayoutView.OptionsSelection.MultiSelect := cbMultiSelectRecords.Checked;
end;

procedure TfrmMain.cbRecordCaptionsClick(Sender: TObject);
begin
  LayoutView.OptionsView.RecordCaption.Visible := cbRecordCaptions.Checked;
end;

procedure TfrmMain.FormCreate(Sender: TObject);

  function GetDir(const AFileName: string; ALevelUp: Integer): string;
  var
    I: Integer;
  begin
    Result := AFileName;
    for I := 1 to ALevelUp do
      Result := ExtractFileDir(Result);
  end;

begin
  Inc(FLockCount);
  try
    tsAutoPitchAngle.Checked := LayoutView.OptionsView.CarouselMode.AutoPitchAngle;
    cbInterpolationMode.ItemIndex := Ord(LayoutView.OptionsView.CarouselMode.InterpolationMode);
    tbPitchAngle.Position := Trunc(LayoutView.OptionsView.CarouselMode.PitchAngle);
    tbRollAngle.Position := Trunc(LayoutView.OptionsView.CarouselMode.RollAngle);
    tbRecordCount.Position := LayoutView.OptionsView.CarouselMode.RecordCount;
    tbBackgroundAlphaLevel.Position := LayoutView.OptionsView.CarouselMode.BackgroundRecordAlphaLevel;
    tbStartRecordScale.Position := LayoutView.OptionsView.CarouselMode.BackgroundRecordStartScale;
    tbEndRecordScale.Position := LayoutView.OptionsView.CarouselMode.BackgroundRecordEndScale;
  finally
    Dec(FLockCount);
  end;
  mdHomes.LoadFromBinaryFile(GetDir(Application.ExeName, 3) + '\Data\Homes.dat');
end;

{$IFDEF EXPRESSBARS}
procedure TfrmMain.PlaceControls;
begin
  DisableAlign;
  lcMain.Align := alNone;
  lbDescription.Align := alNone;
  lbDescription.Top := 100;
  EnableAlign;
  lbDescription.Align := alTop;
  lcMain.Align := alTop;
end;
{$ENDIF}

procedure TfrmMain.CarouselModePropertiesChange(Sender: TObject);
begin
  if FLockCount > 0 then
    Exit;
  Inc(FLockCount);
  try
    LayoutView.OptionsView.CarouselMode.AutoPitchAngle := tsAutoPitchAngle.Checked;
    LayoutView.OptionsView.CarouselMode.InterpolationMode := TdxGPInterpolationMode(cbInterpolationMode.ItemIndex);
    LayoutView.OptionsView.CarouselMode.PitchAngle := tbPitchAngle.Position;
    LayoutView.OptionsView.CarouselMode.RollAngle := tbRollAngle.Position;
    LayoutView.OptionsView.CarouselMode.RecordCount := tbRecordCount.Position;
    LayoutView.OptionsView.CarouselMode.BackgroundRecordAlphaLevel := tbBackgroundAlphaLevel.Position;
    LayoutView.OptionsView.CarouselMode.BackgroundRecordStartScale := tbStartRecordScale.Position;
    LayoutView.OptionsView.CarouselMode.BackgroundRecordEndScale := tbEndRecordScale.Position;
  finally
    Dec(FLockCount);
  end;
  liPitchAngle.Enabled := not tsAutoPitchAngle.Checked;
end;

procedure TfrmMain.FormShow(Sender: TObject);
begin
  inherited;
  LayoutView.DataController.RecNo := LayoutView.DataController.RecordCount div 2;
end;

end.
